src/emulate.c

```diff
@@ -254,100 +254,112 @@ static inline bool insn_is_misaligned(uint32_t pc)
+}
+/* LUI: Load Upper Immediate */
+RVOP(lui)
+{
+    /* LUI is used to build 32-bit constants and uses the U-type format. LUI
+     * places the U-immediate value in the top 20 bits of the destination
+     * register rd, filling in the lowest 12 bits with zeros. The 32-bit
+     * result is sign-extended to 64 bits.
      */
-    _(lui, rv->X[ir->rd] = ir->imm;)
-
-    /* AUIPC (Add Upper Immediate to PC) is used to build pc-relative addresses
-     * and uses the U-type format. AUIPC forms a 32-bit offset from the 20-bit
-     * U-immediate, filling in the lowest 12 bits with zeros, adds this offset
-     * to the address of the AUIPC instruction, then places the result in
-     * register rd.
+    rv->X[rv_reg_zero] = 0;
+    rv->X[ir->rd] = ir->imm;
+    rv->PC += ir->insn_len;
+    rv->csr_cycle++;
+    if (ir->tailcall)
+        return true;
+    nextOP();
+}
+/* AUIPC: Add Upper Immediate to PC */
+RVOP(auipc)
+{
+    /* AUIPC is used to build pc-relative addresses and uses the U-type
+     * format. AUIPC forms a 32-bit offset from the 20-bit U-immediate,
+     * filling in the lowest 12 bits with zeros, adds this offset to the
+     * address of the AUIPC instruction, then places the result in register
+     * rd.
      */
-    _(auipc, rv->X[ir->rd] = ir->imm + rv->PC;)
-
-    /* JAL: Jump and Link
-     * store successor instruction address into rd.
+    rv->X[rv_reg_zero] = 0;
+    rv->X[ir->rd] = ir->imm + rv->PC;
+    rv->PC += ir->insn_len;
+    rv->csr_cycle++;
+    if (ir->tailcall)
+        return true;
+    nextOP();
+}
+/* JAL: Jump and Link */
+RVOP(jal)
+{
+    /* store successor instruction address into rd.
      * add next J imm (offset) to pc.
      */
-    _(jal, {
-        const uint32_t pc = rv->PC;
-        /* Jump */
+    rv->X[rv_reg_zero] = 0;
+    const uint32_t pc = rv->PC;
+    /* Jump */
+    rv->PC += ir->imm;
+    /* link with return address */
+    if (ir->rd)
+        rv->X[ir->rd] = pc + ir->insn_len;
+    /* check instruction misaligned */
+    if (insn_is_misaligned(rv->PC)) {
+        rv->compressed = false;
+        rv_except_insn_misaligned(rv, pc);
+        return false;
+    }
+    /* can branch */
+    rv->csr_cycle++;
+    return true;
+}
+/* JALR: Jump and Link Register */
+RVOP(jalr)
+{
+    /*The indirect jump instruction JALR uses the I-type encoding. The
+     * target address is obtained by adding the sign-extended 12-bit
+     * I-immediate to the register rs1, then setting the least-significant
+     * bit of the result to zero. The address of the instruction following
+     * the jump (pc+4) is written to register rd. Register x0 can be used as
+     * the destination if the result is not required.
+     */
+    rv->X[rv_reg_zero] = 0;
+    const uint32_t pc = rv->PC;
+    /* jump */
+    rv->PC = (rv->X[ir->rs1] + ir->imm) & ~1U;
+    /* link */
+    if (ir->rd)
+        rv->X[ir->rd] = pc + ir->insn_len;
+    /* check instruction misaligned */
+    if (insn_is_misaligned(rv->PC)) {
```

Comment:
Add `unlikely` macro back.

Reply:
You should check the changes line by line to ensure the previous hints to optimizer are still alive.